staker/multi_protocol_staker.go

```diff
@@ -0,0 +1,197 @@
+		fatalErr,
+	)
+	if err != nil {
+		return nil, err
+	}
+	bridge, err := bridgegen.NewIBridge(bridgeAddress, oldStaker.client)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiProtocolStaker{
+		oldStaker:  oldStaker,
+		boldStaker: nil,
+		bridge:     bridge,
+	}, nil
+}
+
+func (m *MultiProtocolStaker) IsWhitelisted(ctx context.Context) (bool, error) {
+	return false, nil
+}
+
+func (m *MultiProtocolStaker) Initialize(ctx context.Context) error {
+	boldActive, rollupAddress, err := m.isBoldActive(ctx)
+	if err != nil {
+		return err
+	}
+	if boldActive {
+		log.Info("BOLD protocol is active, initializing BOLD staker")
+		boldStaker, err := m.setupBoldStaker(ctx, rollupAddress)
+		if err != nil {
+			return err
+		}
+		m.boldStaker = boldStaker
+		return m.boldStaker.Initialize(ctx)
+	}
+	log.Info("BOLD protocol not detected on startup, using old staker until upgrade")
+	return m.oldStaker.Initialize(ctx)
+}
+
+func (m *MultiProtocolStaker) Start(ctxIn context.Context) {
+	m.StopWaiter.Start(ctxIn, m)
+	if m.oldStaker.Strategy() != WatchtowerStrategy {
+		m.oldStaker.wallet.Start(ctxIn)
+	}
+	if m.boldStaker != nil {
+		log.Info("Starting BOLD staker")
+		m.boldStaker.Start(ctxIn)
+		m.StopOnly()
+	} else {
+		log.Info("Starting pre-BOLD staker")
+		m.oldStaker.Start(ctxIn)
+		stakerSwitchInterval := time.Second * time.Duration(m.oldStaker.config.BOLD.CheckStakerSwitchIntervalSeconds)
+		m.CallIteratively(func(ctx context.Context) time.Duration {
+			switchedToBoldProtocol, err := m.checkAndSwitchToBoldStaker(ctxIn)
+			if err != nil {
+				log.Error("staker: error in checking switch to bold staker", "err", err)
+				return stakerSwitchInterval
+			}
+			if switchedToBoldProtocol {
+				log.Info("Detected BOLD protocol upgrade, stopping old staker and starting BOLD staker")
+				// Ready to stop the old staker.
+				m.oldStaker.StopOnly()
+				m.StopOnly()
+			}
+			return stakerSwitchInterval
+		})
+	}
+}
+
+func (m *MultiProtocolStaker) isBoldActive(ctx context.Context) (bool, common.Address, error) {
+	var addr common.Address
+	if !m.oldStaker.config.BOLD.Enable {
+		return false, addr, nil
+	}
+	callOpts := m.oldStaker.getCallOpts(ctx)
+	rollupAddress, err := m.bridge.Rollup(callOpts)
+	if err != nil {
+		return false, addr, err
+	}
+	userLogic, err := boldrollup.NewRollupUserLogic(rollupAddress, m.oldStaker.client)
+	if err != nil {
+		return false, addr, err
+	}
+	_, err = userLogic.ChallengeGracePeriodBlocks(callOpts)
+	// ChallengeGracePeriodBlocks only exists in the BOLD rollup contracts.
+	return err == nil, rollupAddress, nil
+}
+
+func (m *MultiProtocolStaker) checkAndSwitchToBoldStaker(ctx context.Context) (bool, error) {
+	shouldSwitch, rollupAddress, err := m.isBoldActive(ctx)
+	if err != nil {
+		return false, err
+	}
+	if !shouldSwitch {
+		return false, nil
+	}
+	boldStaker, err := m.setupBoldStaker(ctx, rollupAddress)
+	if err != nil {
+		return false, err
+	}
+	if err = boldStaker.Initialize(ctx); err != nil {
```

Comment:
boldStaker.Initilize calls blockValidator.InitAssumeValid, which will fail if called after start.

Reply:
Added a check to only call InitAssumeValid if blockValidator is not started.